src/System.CommandLine.Benchmarks/CommandLine/Perf_Parser_Directives.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.CommandLine.Builder;
+using System.CommandLine.Invocation;
+using BenchmarkDotNet.Attributes;
+
+namespace System.CommandLine.Benchmarks.CommandLine
+{
+    /// <summary>
+    /// Measures the performance directives parsing.
+    /// </summary>
+    [BenchmarkCategory(Categories.CommandLine)]
+    public class Perf_Parser_Directives
+    {
+        private Parser _testParser;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            var option = new Option("-opt");
+
+            _testParser =
+                new CommandLineBuilder()
+                    .AddOption(option)
+                    .UseParseDirective()
+                    .Build();
+        }
+
+        [Params(
+            "[directive1] -opt",
+            "[directive1] [directive2] -opt",
+            "[directive1:1] [directive2:2] -opt",
+            "[directive1] [directive2] [directive2] -opt",
+            "[directive1:1] [directive2:2] [directive2:3] -opt"
+        )]
+        public string TestCmdArgs;
+
+        [Benchmark]
+        public IDirectiveCollection ParseDirectives()
+            => _testParser.Parse(TestCmdArgs).Directives;
+
+        [Benchmark]
+        public string ParseDirectives_Diagram()
+            => _testParser.Parse(TestCmdArgs).Diagram();
```

Comment:
Do you know what the time split between `Parse()` and `Diagram()` is? A quick glance at the code doesn't suggest that `Diagram()` will be cheaper than `Parse()` - I ask because when you mix things together like this you can run the risk of losing regressions in the smaller part beacuse the overall negative contribution isn't big enough. I took a crack at a sample for how this might work [here](https://gist.github.com/billwert/cef0e7eb0156d9db92ca1bf83c6c028b). Sadly `ToString` on a `ParseResult` just calls `Diagram`, so perhaps you can make a small wrapper class which holds the `ParseResult` and has a `ToString` suitable for benchmark naming on it?

Reply:
Fixed. Parsing is now being done separately at setup stage.
src/System.CommandLine.Benchmarks/CommandLine/Perf_Parser_ParseResult.cs

Introduced `BdnParam<T>` wrapper class to substitute `T.ToString()` with a custom comment.
src/System.CommandLine.Benchmarks/Helpers/BdnParam.cs